refinery/tool_manager/utils.py

```diff
@@ -11,3 +17,132 @@ class AdminFieldPopulator(admin.ModelAdmin):
     def __init__(self, model, admin_site):
         super(AdminFieldPopulator, self).__init__(model, admin_site)
         self.list_display = [field.name for field in model._meta.fields]
+
+
+def create_tool_definition_from_workflow(workflow_data):
+    """
+    :param workflow_data: dict of annotated data coming from a Galaxy workflow
+    """
+
+    file_relationship = create_nesting(workflow_data["annotation"])
+
+    tool_definition = ToolDefinitionFactory(
+        name=workflow_data["name"],
+        description=workflow_data["annotation"]["description"],
+        tool_type="WORKFLOW",
+        file_relationship=file_relationship
+    )
+
+    for output_file in workflow_data["annotation"]["output_files"]:
+        tool_definition.output_files.add(
+            OutputFileFactory(
+                name=output_file["name"],
+                description=output_file["description"],
+                filetype=FileType.objects.get(
+                    name=output_file["filetype"]["name"]
+                )
+            )
+        )
+
+    # TODO: figure out Parameter/GalaxyParameter stuff
+
+
+def create_nesting(d, fr_store=None):
+    """
+    :param d: dict to act recursively upon to build the proper
+    FileRelationship structure
+
+    :param fr_store: initially `None`, but becomes a populated list upon
+    recursive calls to create_nesting. Allows for temp.
+    storage of FileRelationship objs created in this method.
+    Is necessary due to the fact that we cannot properly create the M2M
+    relations between FileRelationships until we have created the
+    "bottom-most" one.
+
+    :return: The top-most FileRelationship object
+    """
+
+    if not fr_store:
+        fr_store = []
```

Comment:
Each recursive call appends the next level `file_relationship` to the`fr_store` list. So no overwriting is done, since each recursive call to `create_nesting` has an updated `fr_store` list.

Each function call does have its own scope, but to properly link each `file_relationship` in the nesting, we need to know about every one that is above the bottom-most one when we reach that point.